consensus/consensus-types/src/timeout_certificate.rs

```diff
@@ -57,16 +67,27 @@ impl TimeoutCertificate {
         self.timeout.round()
     }
 
-    /// Returns the signatures certifying the round
-    pub fn signatures(&self) -> &BTreeMap<Author, Ed25519Signature> {
-        &self.signatures
+    /// The highest hqc round of the 2f+1 participants
+    pub fn hqc_round(&self) -> Round {
+        self.timeout.hqc_round()
     }
 
-    pub fn add_signature(&mut self, author: Author, signature: Ed25519Signature) {
-        self.signatures.entry(author).or_insert(signature);
+    /// Returns the signatures certifying the round
+    pub fn signers(&self) -> impl Iterator<Item = &Author> {
+        self.signatures.iter().map(|(k, _)| k)
     }
 
-    pub fn remove_signature(&mut self, author: Author) {
-        self.signatures.remove(&author);
+    /// Add a new timeout message from author
+    pub fn add(&mut self, author: Author, timeout: Timeout, signature: Ed25519Signature) {
+        assert_eq!(
+            self.timeout.round(),
+            timeout.round(),
+            "Timeout should have the same round as TimeoutCert"
+        );
+        let hqc_round = timeout.hqc_round();
+        if timeout.hqc_round() > self.timeout.hqc_round() {
+            self.timeout = timeout;
+        }
+        self.signatures.insert(author, (hqc_round, signature));
```

Comment:
Sorry for the petty comments, I hope they are useful...
1. you are not checking epoch in the assert.
2. why don't you verify the signature here? one bad signature (by bad party) can make the verify Fn return error (and possibly prevent liveness?)

Reply:
the structure we organize the code is layered.
- at network layer we verify messages (all the invariants and signatures, see the verify function) and pass to epoch manager 
- at epoch manager layer we route messages from different epoch and pass to round manager
- at round manager layer we route messages from different rounds, so anything reaches here should have the same epoch/round, the assert is more served as debug check as I develop but additional assert doesn't hurt much